### DISCUSSION TITLE
Roadmap for 24.11

### DISCUSSION THREAD

**Discussion by fballiano** [MahoCommerce/maho#21] [Roadmaps]
Task                                                                                                                                                  | Status
------------------------------------------------------------------------------------------ | ------
Fix some CodeQL security issues                                                                                                  | ✅
Replace jscolor with Simonwep/pickr? No, jscolor is not used, removed                                   | ✅
Replace js calendar by Mihai Bazon (last release in 2005) with flatpickr (MIT, modern)          | ✅
RWD: remove sharing from product page                                                                                     | ✅
RWD: Remove jquery.                                                                                                                      | ✅
[Simplify persistent templates](#34)                                                                                             | ✅
Possibility to disable MageTag for frontend                                                                                  | ✅
[Migration guide](https://mahocommerce.com/migrate-to-maho)                                            | ✅
Continue cleanup process                                                                                                              | ✅
[EAV attribute editor](https://github.com/MahoCommerce/maho/pull/25)                                | [moved to next release](#52)
Moving "awesome list" of modules to maho's website                                                                | [moved to next release](#52)
Replace extjs tree? eg: cheton/infinite-tree                                                                                  | [moved to next release](#52)

**Want more?**  
Consider [getting involved](https://mahocommerce.com/community/get-involved) or [sponsoring](https://mahocommerce.com/sponsors) ❤️

**Comment by Sdfendor** [comment]
I would like to propose to get rid of leading underscores to show visibility in classes for methods and member variables. I'm not sure if this should be tackled in one go or in parts though.

**Reply by fballiano** [reply]
Hi @Sdfendor, it's true that the _ is not used anymore to indicate property/method visibility, but, wouldn't it break every custom inherited class?
I'm pro changing and moving forward but would this change give us such a value improvement to justify breaking every custom code out there?

**Reply by Sdfendor** [reply]
It would. I'm still not sure about the stance on breaking things in this project, even major things tbh. In our current OM compatible codebase (as reference) we removed the leading underscores but had to keep them in place for all the overrides of course. We put additional work in to mark all the overriden core methods with the Override attribute just to make clear to the team where any remaining underscore is coming from. The attribute has other benefits of course.
Furthermore it would be great to be able to not have to "switch context" when working with Maho code in the future. Maybe it's too early for this, maybe we could start by not adding any more new underscore methods and member variables at the risk of a uneven codebase for the foreseeable future.

**Reply by fballiano** [reply]
I also think it's probably too early because we wouldn't be able to use probably no module at all which is not the best.

Also check this commit https://github.com/MahoCommerce/maho/commit/89753481ecfaec16ff78992ef9a8c3141644d9a8 ;-)

**Reply by Sdfendor** [reply]
Oh really nice 👍. 
I'm unfortunately not yet really familiar with the current codebase and all the changes made since the project has started but I'm getting there.

**Reply by fballiano** [reply]
I'll try to document as much as possible in the first release notes that should come in the next few days

**Reply by tmewes** [reply]
If I may add my opinion briefly, I would like to get it off my chest, even though most of the things have already been said.

I'm also in favor of getting rid of the underscores (that was to be expected :)). The only reasons why I haven't t yet started preparing a PR for this are (unfortunately) time and that the direction of the project is not yet completely clear. In other words, how such devastating changes are dealt with. In principle, there would again be a note in the release and a migration guide or (as soon as they are available) further commands in the respective migration script.

I also do not think that this change has to be forced for the first "real" releases. We have time for this. Probably until the migration scripts mentioned for OM, for example, are available. This would allow us to create an automatism that also adapts the custom code, and this change is no longer as 'dangerous' as it is fixed just in the core and all users are 'on their own'.

It's really nice to see a breath of fresh air coming in here, it's great fun and opens up a great future.
There are already so many profound changes in the first release. Let us all (including us @Sdfendor) get to grips with them and use them. :)

**Reply by fballiano** [reply]
My vision is to
- "drop the dead": remove ancient stuff, scriptaculous, prototype, unused modules (like the backup, sendfriend etc)
- introduce new modern features (enhanced grid, eav attribute editor, dynamic categories) that are interesting out-of-the-box for the users (also because many modules out there became so outdated that we have to start from scratch)
- make the phtmls more modern (there are a million of redundant DOM nodes) -> CSS too
- the new base theme
- make it more modern for developers (the laravel-artisan approach)

But, if we have to break compatibility it should be for good reason, I love a clean code but IMHO, until OM is alive we should try to have some sort of compatibility because it allows us to bring their PRs into our code easily, and this is absolutely necessary.

Removing the underscore could look good for developers but it destroys almost all compatibilities and I'm not sure if it's worth it.

**Reply by fballiano** [reply]
For sure we should avoid using underscores for new code, and we could remove it from private properties/methods, but for protected/public... mmmmmmmmmm I think the risk is too high for the provided value.

**Comment by toejam34** [comment]
Could I suggest importers? Until you have these you wont have a userbase to carry things forward and drive further requirements. 

Magento 1.9.4.5 --> MahoCommerce
OpenMage 19.4.5 --> MahoCommerce
OpenMage 20.x --> MahoCommerce

**Reply by fballiano** [reply]
Hi @toejam34, do you mean an upgrade-guide? because there should be a need for importing the database because it's absolutely compatible. But I'm already working on an upgrade guide, but yes it should be in the roadmap.

**Reply by toejam34** [reply]
I didnt realise it was drop in compatible! An upgrade guide would be great!

**Reply by fballiano** [reply]
:-) upgrade guide absolutely necessary, right now I've already pushed a "./maho health-check" that will kinda guide you during the update, although it doesn't replace the upgrade guide completely.

**Comment by toejam34** [comment]
Id like to see a central repository for plugins/add-ons. 

At present I find the other systems rather disjointed for finding addons. If I want a plugin for say Cloudflare turnstile I visit one GitHub , if I want something to do with Sorting of a customer I visit another site. 

It might be outside the scope of 24.11.0 but it would be superb if we could get a central marketplace for free and paid items as I feel this is one of the major strengths of the Adobe implementation - you can easily see what is available as an addon.

**Reply by fballiano** [reply]
Hi @toejam34, at the moment the resource to find module is https://github.com/fballiano/awesome-openmage, which I hand-picked, would it be ok if we start by moving it inside maho's website? I know a store has a search funcionality but... it would be a first step, and we can leave a contact to suggest new modules to add

**Reply by toejam34** [reply]
I think that is a solid start.

For me its all about and end to end experience. As a semi literate cart administrator I can see how github hangs items together. However some people are not read into the github thing and want a AIO for addons etc. Something where a vendor can submit to. 

I think this is one of the crucial things that stops widespread adoption.

**Reply by fballiano** [reply]
At the moment I wouldn't build a real store because I don't think many vendors would submit anything, but moving the "awesome list" to the website makes a lot of sense, I'll add it to the roadmap (should be quite easy actually)

**Reply by toejam34** [reply]
I suspect if you did build a real store and a certain fballiano started to populate it with addons that offer more than just the core product more and more would become interested.

Its a bit of a catch 22. Need devs to release items for a store, need a people to use Maho to interest devs in releasing.

**Reply by fballiano** [reply]
heheheh let's start with a first step because we're already developing like crazy and it wouldn't be possible to create a real module store at the moment :-)

**Reply by toejam34** [reply]
This is your party! You do whatever you see fit!
